### PR TITLE
[@types/react] make optional createContext defaultValue argument

### LINKED ISSUE #37023: [@types/react] createContext – An argument for 'defaultValue' was not provided.
Is there any reason of why the argument `defaultValue` is not optional here?

```tsx
import * as React from 'react';
import * as firebase from 'firebase/app';

interface AuthProps {
  children: any;
}
interface AuthState {
  user: firebase.User | null;
}
interface AuthContextProps extends AuthState {
  updateUser(user: firebase.User | null): void;
}

export const AuthContext = React.createContext<AuthContextProps>(); // <-- ERROR HERE!

// export const AuthConsumer = AuthContext.Consumer;
export const AuthProvider = (props: AuthProps) => {
  const [state, setState] = React.useState<AuthState>({
    user: null,
  });

  const updateUser = (user: firebase.User | null) => setState({ user });

  return (
    <AuthContext.Provider
      value={{
        updateUser,
        user: state.user,
      }}
    >
      {props.children}
    </AuthContext.Provider>
  );
};

```

Docs url: https://reactjs.org/docs/context.html#reactcreatecontext

> The defaultValue argument is only used when a component does not have a matching Provider above it in the tree.

I have a matching Provider above in the tree, so I should be able to create a context without set a default value or I'm wrong?

### PR DESCRIPTION
**Pull request by exwyezed**
This PR fixes #37023

### CODE REVIEW

File: types/react/index.d.ts [hunk 1]
```diff
@@ -334,7 +334,7 @@ declare namespace React {
         displayName?: string;
     }
     function createContext<T>(
-        defaultValue: T,
+        defaultValue?: T,
```
Comment (eps1lon) [inline types/react/index.d.ts]: This means you could write `createContext<object>();` and get no compile errors when accessing the context unguarded.

What we want is probably an overload. Something like
```ts
function createContext<T>(): Context<T | undefined>;
```

**Review by eps1lon — CHANGES_REQUESTED** [review]

File: types/react/index.d.ts (hunk 1 above)
Comment (alfonmga) [reply types/react/index.d.ts]: Looks good to me, should I apply the suggested change `function createContext<T>(): Context<T | undefined>;` and we can LGTM?

Comment (eps1lon) [reply types/react/index.d.ts]: Sure. But don't forget to add a test.

**Comment by alfonmga** [comment]
I'm not 100% sure about this PR, I'm going to close this and let others decide because maybe I'm wrong as I don't have enough experience with TypeScript.

**Pull request closed without merging**